[1단계]  블랙잭

RoadtoS7 opened the pull request:
안녕하세요 리뷰어님!
블랙잭을 구현해보았습니다!!
반복적으로 사용자의 입력을 받는 부분과 랜덤으로 카드를 뽑는 부분에서 테스트 가능한 부분과 불가능한 부분을 분리하는데 매우 어렵더라구요..ㅠㅠㅠ
해결방법 중 하나로 delegate 패턴을 사용해보았는데 `BlackjackGame` 인스턴스에대한 결합도가 커진 것 같습니다.. 쉽지 않네요.. 😭 
피드백 부탁드립니다 감사합니다!!
p.s. 컨벤션에 어긋나거나, swift 적인 방법을 활용했을 때 더 효율적으로 해결될 수 있는 것이 있다면 피드백 부탁드립니다!

RoadtoS7 committed:
initial commit

RoadtoS7 committed:
feat: PlayerName이 콤마를 입력으로 받으면 nil을 반환하는지 체크하는 테스트 케이스 추가

RoadtoS7 committed:
feat: PlayerName이 유닛 테스트를 통과하도록 구현

RoadtoS7 committed:
feat: PlayerName 리팩터링

RoadtoS7 committed:
feat: 카드 한장을 나타내는 Card 구현

RoadtoS7 committed:
feat: 트럼프 카드를 나타내는 TrumpCards 구현

RoadtoS7 committed:
feat: CardDeck에 카드를 추가하는 메서드 유닛 테스트 케이스 추가

RoadtoS7 committed:
feat: CardDeck에 카드를 추가하는 메서드 append 구현

RoadtoS7 committed:
feat: 게임 규칙에 따라서 카드점수를 계산하는 CardScoreRule 유닛 테스트 작성

RoadtoS7 committed:
feat: CardDeckScore 구현

RoadtoS7 committed:
feat: CardDeckScore에서 ace가 1로 계산되는 테스트 케이스 추가

RoadtoS7 committed:
feat: 사용자에게 참여자 이름을 받는 InputView 메서드 구현

RoadtoS7 committed:
feat: 사용자로부터 입력받은 이름을 PlayerName으로 파싱하는 PlayerNameParser 구현

RoadtoS7 committed:
feat: PlayerNameParser 구현

RoadtoS7 committed:
feat: PlayerNameParser 리팩터링

RoadtoS7 committed:
feat: PlayerNameParser에서 에러를 throw하는 경우 유닛 테스트 추가

RoadtoS7 committed:
feat: PlayerNameParser에서 에러를 throw하도록 구현

RoadtoS7 committed:
feat: PlayerNameParser 리팩터링

RoadtoS7 committed:
feat: OutputView 생성 및 error 출력 메서드 구현

RoadtoS7 committed:
feat: main에서 사용자로부터 이름입력받아 PlayerName으로 파싱하는 로직 구현

RoadtoS7 committed:
feat: CardPickStrategy 구현

RoadtoS7 committed:
feat: NonDuplicateCardDistributor가 중복되지 않게 카드를 분배하는지 체크하는 유닛테스트 추가

RoadtoS7 committed:
feat: NonDuplicateCardDistributor 구현

RoadtoS7 committed:
feat: Player.receive(card:) 유닛 테스트 구현

RoadtoS7 committed:
feat: Player.receive(card:) 구현

RoadtoS7 committed:
feat: CardDeck에 존재하는 카드의 합계점수를 계산하는 countScore 유닛테스트 구현

RoadtoS7 committed:
feat: CardDeck.countScore 구현

RoadtoS7 committed:
feat: AnswerParser의 테스트 케이스 구현

- 'y' -> Answer.yes로 파싱한다
- 'n' -> Answer.no로 파싱한다
- ' ' -> error를 throw한다

RoadtoS7 committed:
feat: AnswerParser 내용 구현

RoadtoS7 committed:
feat: 카드 수령여부를 받는 AnswerReaderDelegate 구현

RoadtoS7 committed:
feat: 게임의 각 단계 상황을 출력하는 역할을 하는 GameStateDelegate 구현

RoadtoS7 committed:
feat: Player와 Player가 소지하는 카드를 출력 포맷으로 만드는 PlayerFormatter 구현

RoadtoS7 committed:
feat: Player에게 카드 2장 나누어준 것, Player 상태 출력 메서드 및 점수 출력  메서드 구현

RoadtoS7 committed:
feat: Blackjack은 게임이 시작전에 모든 참가자에게 카드 2장씩 배부하는지 유닛 테스트 구현

RoadtoS7 committed:
feat: Blackjack게임이 초기화될 때 모든 참가자에게 카드 2장씩 배부하도록 구현

RoadtoS7 committed:
feat: 게임 결과를 담는 BlackjackScore 구현

RoadtoS7 committed:
feat: Player가 게임결과를 계산하는 countScore() 메서드 구현

RoadtoS7 committed:
feat: start메서드를 호출하면 player는 yes라고 대답한 횟수만큼 카드를 수령한다

RoadtoS7 committed:
feat: BlackjackScore를 출력 string 형식으로 포맷하는 BlackjackScoreFomatter 구현

RoadtoS7 committed:
feat: main 로직 구현

RoadtoS7 committed:
merge conflict

zdodev commented on Blackjack/AnswerReadDelegate.swift:
```diff
@@ -0,0 +1,20 @@
+    func readAnswer(player: Player) throws -> Answer
+}
+
+struct BlackjackAnswerReader: ReadAnswerDelegate {
```
플레이어의 반복적인 입력을 분리해보셨네요!
`BlackjackGame` 타입이 이런 책임까지 가지고 있는 것보다 이런식으로 분리하게 되면 `BlackjackGame` 역할이 더욱 명확지지 않을까 싶습니다!👍

zdodev commented on Blackjack/BlackjackGame.swift:
```diff
@@ -0,0 +1,60 @@
+struct BlackjackGame {
+    let players: [Player]
+    let cardDistributor: CardDistributor
+    let readAnswerDelegate: ReadAnswerDelegate
```
더 나은 접근 레벨을 선언해도 좋을 것 같습니다!

RoadtoS7 replied:
[34d97ba](https://github.com/next-step/swift-blackjack/pull/8/commits/34d97babeb824c7f020a4f772cb7fe2235fc4984) 커밋에서 반영했습니다! 피드백 감사합니다!~

zdodev commented on Blackjack/BlackjackGame.swift:
```diff
@@ -0,0 +1,60 @@
+    let players: [Player]
+    let cardDistributor: CardDistributor
+    let readAnswerDelegate: ReadAnswerDelegate
+    let gameStateDelegate: GameStateDelegate?
```
`gameStateDelegate` 타입을 옵셔널로 선언하신 이유가 있을까요?

RoadtoS7 replied:
gameStateDelegate는 게임의 중간 상황을 외부에 알릴 때 사용하는 객체라고 생각했어요!
gameStateDelegate 인스턴스가 없어서 `BlackjackGame` 의 중간 징행 상황을 외부에 알릴 수 없더라도 `BlackjackGame`을 실행할 수는 있어야 한다고 생각하여 optional로 구현했습니다!

zdodev replied:
좋습니다!

zdodev commented on Blackjack/BlackjackGame.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    private func askAboutCard(to player: Player) throws {
+    while true {
```
`while` 문의 들여쓰기 수준을 맞춰볼 수 있을까요?

RoadtoS7 replied:
[d54dab4](https://github.com/next-step/swift-blackjack/pull/8/commits/d54dab4c5ad69802256c5c5f2600d3cf47765ee5) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/BlackjackScore.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+struct BlackjackScore {
+    let whos: Player
```
`player` 라는 이름은 어떤가요?🙃

RoadtoS7 replied:
오 `player` 가 더 직관적이네요! [7b780fa](https://github.com/next-step/swift-blackjack/pull/8/commits/7b780fa2322db01b6e5b65d609230317e9105ef3) 커밋에서 반영했습니다. 피드백 감사합니다!!

zdodev commented on Blackjack/Card.swift:
```diff
@@ -0,0 +1,33 @@
+    let symbol: CardSymbol
+}
+
+enum CardId: CaseIterable {
```
1. `CardId` 대신에 카드 게임에서 사용하는 `Rank` 라는 용어는 어떨까요? 보다 명확하지 않을까 싶습니다! 의견이 궁금합니다!
2. `CardId` 를 외부에 정의하신 이유가 궁금합니다! `Card` 타입에 중첩 타입으로 구현해보는 것은 어떨까요? 의견이 궁금합니다!

RoadtoS7 replied:
> CardId 대신에 카드 게임에서 사용하는 Rank 라는 용어는 어떨까요? 보다 명확하지 않을까 싶습니다! 의견이 궁금합니다!

도메인에 맞는 용어를 사용하는 것이 코드를 파악할 때 더 도움된다고 생각합니다! 
[598c116](https://github.com/next-step/swift-blackjack/pull/8/commits/598c116cd6c708ed6b4d9187908765728af69304) 커밋에서 Rank로 수정했습니다!

추가적으로 하트, 스페이드와 같은 카드의 문양도 블랙잭에서 Suit로 불린다고 하여, 기존의 CardSymbol에서 Suit로 이름을 수정했습니다! ([4c669fd](https://github.com/next-step/swift-blackjack/pull/8/commits/4c669fd04af51eff0540f9ba7a7408e9d9688d0a) 커밋)

> CardId 를 외부에 정의하신 이유가 궁금합니다! Card 타입에 중첩 타입으로 구현해보는 것은 어떨까요? 의견이 궁금합니다!

저도 CardId가 Card 프로퍼티의 타입이고, Card 구성요소니까 중첩 타입이 더 좋은 것 같습니다! 
[b8def00](https://github.com/next-step/swift-blackjack/pull/8/commits/b8def0090197e9800c00aedfa84cba5eeccd1c56) 커밋에서 반영했습니다!

피드백 감사합니다!!~

zdodev commented on Blackjack/Card.swift:
```diff
@@ -0,0 +1,33 @@
+    case king
+}
+
+enum CardSymbol: CaseIterable {
```
`CardSymbol` 를 외부에 정의하신 이유가 궁금합니다! `Card` 타입에 중첩 타입으로 구현해보는 것은 어떨까요? 의견이 궁금합니다!

RoadtoS7 replied:
[b8def00](https://github.com/next-step/swift-blackjack/pull/8/commits/b8def0090197e9800c00aedfa84cba5eeccd1c56) 커밋에서 반영했습니다! 피드백 감사합니다!!~

zdodev commented on Blackjack/CardDeck.swift:
```diff
@@ -0,0 +1,28 @@
+
+class BlackjackCardDeck: CardDeck {
+    private(set) var cards: [Card] = []
+    var cardScoreRule: CardScoreRule.Type = BlackjackScoreRule.self
```
`var` 로 선언하신 이유가 있을까요?

RoadtoS7 replied:
`var` 로 선언될 필요가 없는 프로퍼티 이네요😲 
[0bd043c](https://github.com/next-step/swift-blackjack/pull/8/commits/0bd043c1d47dfffd5a1359f85114f094fc48ffec) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardDistributor.swift:
```diff
@@ -0,0 +1,30 @@
+}
+
+class NonDuplicateCardDistributor: CardDistributor {
+    private(set) var cards: [Card]
```
블랙잭 덱의 카드를 배열로 관리하도록 구현해보셨네요.
블랙잭 게임 특성 상 덱 내 중복된 카드가 존재하지 않는데, 중복이 가능한 배열보다는 중복을 허용하지 않는 타입으로 관리하는 것은 어떨까요?
의견이 궁금합니다! 지금도 좋습니다!👍

RoadtoS7 replied:
NonDuplicateCardDistributor라는 이름에도 중복되지 않는다고 명시가 되어있으니까 중복을 허용하지 않는 타입이 더 좋다고 생각합니다!
[3aa3e42](https://github.com/next-step/swift-blackjack/pull/8/commits/3aa3e42a6116358baf638225950fe297d705480c) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardDistributor.swift:
```diff
@@ -0,0 +1,30 @@
+
+class NonDuplicateCardDistributor: CardDistributor {
+    private(set) var cards: [Card]
+    let cardPickStrategy: CardPickStrategy
```
`internal` 레벨로 선언하신 이유가 있을까요?

RoadtoS7 replied:
특별히 이유는 없습니다..!
혹시 final class로 선언하는 것이 더 좋을까요..?

zdodev replied:
이 부분은 `cardPickStrategy` 선언을 얘기드린거 였습니다!
class 부분은 상속하지 않는다면 final 키워드를 사용하는 것이 더 좋을 것 같습니다!

RoadtoS7 replied:
internal로 선언한 이유가 없어서 [8d62d31](https://github.com/next-step/swift-blackjack/pull/8/commits/8d62d310d5b7646b235a209c46a6df3a7660acb1)
 커밋에서  private으로 수정했습니다! 
피드백 감사합니다!!

zdodev commented on Blackjack/CardDistributor.swift:
```diff
@@ -0,0 +1,30 @@
+    let cardPickStrategy: CardPickStrategy
+    
+    init(cards: [Card], cardPickStrategy: CardPickStrategy) {
+        self.cards = Array(cards)
```
`cards` 를 `Array()` 로 한번 감싸신 이유가 궁금합니다!

RoadtoS7 replied:
처음 CardDistributor를 구현할 때 initializer에서 `cards` 인자를 `Collection<Card>`으로 받도록 구현했었는데, 이 값을 `cards` 프로퍼티에 할당하려면 Array로 캐스팅이 필요했었어서 그때 코드가 남은 것 같습니다 😅
[3aa3e42](https://github.com/next-step/swift-blackjack/pull/8/commits/3aa3e42a6116358baf638225950fe297d705480c) 커밋에서 수정했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardPickStrategy.swift:
```diff
@@ -0,0 +1,21 @@
+        let range = cards.startIndex..<cards.endIndex
+        let randomIndex = Int.random(in: range)
+        return cards[randomIndex]
+    }
```
인덱스를 통해 랜덤 카드를 구해보셨네요!
`cards` 의 `randomElement()` 메서드 대신에 인덱스를 통해 구하신 이유가 궁금합니다!

RoadtoS7 replied:
엇..! `randomElement()`이라는 좋은 메서드가 있군요..! 몰랐습니다😱
[3aa3e42](https://github.com/next-step/swift-blackjack/pull/8/commits/3aa3e42a6116358baf638225950fe297d705480c) 커밋에서 수정했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,62 @@
+    
+    static func countScore(cardDeck: CardDeck) -> Int {
+        let sumWhereAceOne = totalScore(cardDeck: cardDeck, isAceEleven: false)
+        let sumeWhereAceEleven = totalScore(cardDeck: cardDeck, isAceEleven: true)
```
오타가 있어보입니다!

RoadtoS7 replied:
[d803e46](https://github.com/next-step/swift-blackjack/pull/8/commits/d803e46cfd22ee1b3d2f4da46b7dca5f91511220)  커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+    
+    private static func bestScore(sumWhereAceOne: Int, sumWhereAceEleven: Int) -> Int {
```
`best` 대신 `최적` 이라는 의미를 담는 용어는 어떤가요? 항상 가장 높은 값을 반환한다는 의미로 생각될 수 있지 않을까 싶었습니다!
의견이 궁금합니다! `best` 도 좋습니다!👍

RoadtoS7 replied:
말씀을 듣고 보니 가장 높은 값을 리턴하는 것으로 인식될 수 있겠네요!! [d5f3f1c](https://github.com/next-step/swift-blackjack/pull/8/commits/d5f3f1c7468940f10f075d1c38cdc4b0da34dbbb) 커밋에서 `idealScore()`로 변경했습니다! 피드백 감사합니다!!

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,62 @@
+        
+        let minGap = min(gapWhereAceOne, gapWhereAceEleven)
+        
+        if minGap == gapWhereAceOne {
```
나중에 코드를 봤을 때, `minGap == gapWhereAceOne` 이 어떤 의미의 조건인지 파악하기 쉽지 않아보입니다!
어떤 의미인지 나타낼 수 있도록 프로퍼티나 메서드를 사용해보는 것은 어떨까요?

RoadtoS7 replied:
피드백 감사합니다!!
[d803e46](https://github.com/next-step/swift-blackjack/pull/8/commits/d803e46cfd22ee1b3d2f4da46b7dca5f91511220) 커밋에서 별도의 메서드로 빼내는 방식으로 수정해보았습니다!  이 로직에 대해 의견 여쭙고 싶습니다..!

zdodev replied:
피드백으로 남기겠습니다!

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,62 @@
+        let minGap = min(gapWhereAceOne, gapWhereAceEleven)
+        
+        if minGap == gapWhereAceOne {
+            if gapWhereAceOne >= 0 { return sumWhereAceOne }
```
현재 코드 상 들여쓰기 단계는 1단계이지만, if문 안에 if문이 작성되어 있습니다. 추후에 코드가 추가되면 들여쓰기 단계가 늘어날 수 있지 않을까 싶습니다!
이 부분에 들여쓰기 단계가 생기지 않도록 예방해 볼 수 있을까요?

RoadtoS7 replied:
피드백 감사합니다!!
[d803e46](https://github.com/next-step/swift-blackjack/pull/8/commits/d803e46cfd22ee1b3d2f4da46b7dca5f91511220) 커밋에서 로직을 수정해보았습니다! (https://github.com/next-step/swift-blackjack/pull/8#discussion_r881517663 리뷰에 대한 커밋과 동일합니다)

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,62 @@
+            return sumWhereAceEleven
+        }
+        
+        if gapWhereAceEleven >= 0 { return sumWhereAceEleven }
```
나중에 코드를 봤을 때, `gapWhereAceEleven >= 0` 이 어떤 의미인지 파악하기 쉽지 않아보입니다.
어떤 의미인지 나타낼 수 있도록 프로퍼티나 메서드를 사용해보는 것은 어떨까요?

RoadtoS7 replied:
피드백 감사합니다!!
[d803e46](https://github.com/next-step/swift-blackjack/pull/8/commits/d803e46cfd22ee1b3d2f4da46b7dca5f91511220) 커밋에서 로직을 수정해보았습니다! (https://github.com/next-step/swift-blackjack/pull/8#discussion_r881517663 리뷰에 대한 커밋과 동일합니다)

zdodev commented on BlackjackTest/BlackjackGameTest.swift:
```diff
@@ -0,0 +1,66 @@
+        let answerReader = StubAnswerReader()
+        
+        // when
+        _ = BlackjackGame(players: [playerOne, playerTwo],
+                                         cardDistributor: cardDistributor,
+                                         answerReaderDelegate: answerReader)
```
첫 번째 파라미터와 동일하게 나머지 파라미터의 들여쓰기 수준을 맞춰보는 것은 어떨까요?

RoadtoS7 replied:
[b9cd81f](https://github.com/next-step/swift-blackjack/pull/8/commits/b9cd81f9996a7fff3e4b26cad931d073c32fb252) 커밋에서 반영했습니다! 피드백 감사합니다!~

zdodev commented on BlackjackTest/BlackjackGameTest.swift:
```diff
@@ -0,0 +1,66 @@
+        let answerReader = StubAnswerReader()
+        answerReader.answer = [.yes, .yes, .no, .yes, .yes, .yes, .no]
+        
+        let blackjackGame = BlackjackGame(players: [playerOne, playerTwo],
+                                         cardDistributor: cardDistributor,
+                                         answerReaderDelegate: answerReader)
```
첫 번째 파라미터와 동일하게 나머지 파라미터의 들여쓰기 수준을 맞춰보는 것은 어떨까요?

RoadtoS7 replied:
[b9cd81f](https://github.com/next-step/swift-blackjack/pull/8/commits/b9cd81f9996a7fff3e4b26cad931d073c32fb252) 커밋에서 반영했습니다! 피드백 감사합니다!

zdodev commented on BlackjackTest/BlackjackGameTest.swift:
```diff
@@ -0,0 +1,66 @@
+        XCTAssertTrue(playerOne.cardDeck.cards.count == 4)
+        XCTAssertTrue(playerTwo.cardDeck.cards.count == 5)
+    }
+ }
```
공백이 하나 삽입된 것 같습니다!

RoadtoS7 replied:
[352285e](https://github.com/next-step/swift-blackjack/pull/8/commits/352285ed223da7949db2ec1465fed2b4ee88fda4) 커밋에서 반영했습니다! 피드백 감사합니다! 
혹시 이런 추가적인 공백이나 오타 부분을 하이라이팅해서 보여주는 기능이 xcode에 있나요??

zdodev replied:
`swiftlint` 를 통해서 최대한 잡을 수 있지 않을까 싶습니다. 오타의 경우 문법 에러가 아니면 잡기가 쉽지 않지 않을까 싶습니다!
`swiftlint` 를 사용해도 모든 내용을 잡을 수는 없기 때문에 코드 정렬을 위해 `ctrl + i`(Re-Indent) 를 자주 사용하곤 합니다!😁

zdodev commented on BlackjackTest/BlackjackTest.swift:
```diff
@@ -0,0 +1,35 @@
+
+import XCTest
+
+class BlackjackTest: XCTestCase {
```
사용하지 않는 테스트 코드로 보입니다! 제거해보는 것은 어떨까요?

RoadtoS7 replied:
[ab13a94](https://github.com/next-step/swift-blackjack/pull/8/commits/ab13a94057174dcf0f1619e51de410ac380b1972) 커밋에서 반영했습니다! 피드백 감사합니다!!

zdodev commented on BlackjackTest/CardDistributorTest.swift:
```diff
@@ -0,0 +1,33 @@
+
+import XCTest
+
+struct StubCardPickStrategy: CardPickStrategy {
```
`stub` 을 사용해보셨네요.👍

zdodev commented on BlackjackTest/CardDistributorTest.swift:
```diff
@@ -0,0 +1,33 @@
+    func test_NonDuplicateCardDistributor_뽑은카드는_더이상_CardDistributor에서_앞으로_반환할_카드_대상에_포함되지않는다() {
+        // given
+        let firstCard = Card(id: .ace, symbol: .heart)
+        let secondCard = Card(id: .ace, symbol: .diamond)
```
`secondCard` 카드까지 생성하신 이유가 궁금합니다!

RoadtoS7 replied:
카드 집합에 카드를 한장만 두고서 그것을 뽑는 것보다, 여러개를 두는게 테스트가 더 의미있다고 생각했습니다!
근데 피드백 받은 후에 다시 생각해보니,, 카드 한장을 두고 거기서 뽑는 거나, 여러장을 두는 거나 크게 차이가 없는 것 같네요 🧐
[8b25a35](https://github.com/next-step/swift-blackjack/pull/8/commits/8b25a3541f1eedaaa6219f5630e179db0ecfaf16) 커밋에서 firstCard만 두고 사용하도록 수정했습니다! 피드백 감사합니다!!

zdodev commented on BlackjackTest/PlayerNameTest.swift:
```diff
@@ -0,0 +1,14 @@
+import XCTest
+
+class PlayerNameTest: XCTestCase {
+    func test_콤마는_playername으로_가질수없다() {
```
`playername` 이 다른 테스트 메서드 컨벤션과 달라보입니다!

RoadtoS7 replied:
[76dd623](https://github.com/next-step/swift-blackjack/pull/8/commits/76dd623344c2bfea395c6fd8e3efec644cd9e05c) 커밋에서 
`test_콤마를_이용하여_PlayerName을_생성하면_nil을_반환한다()` 이름으로 수정했는데 피드백을 제가 잘 이해한 것이 맞는지 궁금합니다!

zdodev replied:
네 맞습니다! 다른 테스트의 메서드 이름에서 타입명을 사용할 때는 대문자를 사용했는데, `playername` 의 경우는 모두 소문자로 작성되어 있어서 코멘트를 남겨보았습니다!

zdodev commented on BlackjackTest/PlayerTest.swift:
```diff
@@ -0,0 +1,39 @@
+        let score = player.countScore()
+        
+        // then
+        XCTAssertTrue(score == 21)
```
다른 비교 테스트 구문(`XCTAssertEqual`)과 다르게 `XCTAssertTrue` 로 작성해주셨네요. 일정한 방식으로 작성해보는 것은 어떨까요?

RoadtoS7 replied:
오 일정한 방식으로 작성하는 것이 테스트 코드에 일관성이 있어서 더 읽기 편하겠네요!!
[3d53c88](https://github.com/next-step/swift-blackjack/pull/8/commits/3d53c8860d2c4840bc43755b6b2d43788774792c) 커밋에서 반영했습니다! 피드백 감사합니다!!~

zdodev commented on BlackjackTest/TrumpCardsTest.swift:
```diff
@@ -0,0 +1,62 @@
+    func test_TrumpCards는_트럼프카드규칙대로_카드를갖는다() {
+        // given
+        let heartAce = Card(id: .ace, symbol: .heart)
+        let heartTwo = Card(id: .two, symbol: .heart)
```
`heartAce` 프로퍼티를 제외하고는 다른 프로퍼티는 테스트에 사용하고 있지 않아, 나머지 프로퍼티(카드)를 선언한 이유와 테스트 의도를 알기가 어려워보입니다.
테스트 의도를 알 수 있도록 테스트 코드를 변경해 볼 수 있을까요?

RoadtoS7 replied:
엇.. 구현하다가 미완으로 남겨두었었네요😅
[c2401cd](https://github.com/next-step/swift-blackjack/pull/8/commits/c2401cdcc6459f407e171f83486b89ad5d351086) 커밋에서 마저 작성해보았습니다! 피드백 부탁드립니다 !!

zdodev commented on README.md:
```diff
@@ -1,3 +1,25 @@
 ### Effective Swift
 NEXTSTEP with yagom-academy
+
+## 프로그램 요구사항 분석
```
👍👍

zdodev reviewed (changes requested):
안녕하세요! @RoadtoS7 

블랙잭 1단계 잘 구현해주셨습니다!
원시값 포장과 책임, 역할이 잘 분리되어 읽기 수월했습니다!
코드에 고민한 흔적이 많이 보였습니다!👍

코멘트를 좀 남겨보았습니다!
의견을 나눠보면 좋을 것 같습니다!😁
모든 코멘트에 다 개선 및 수정을 하지 않으셔도 됩니다! 중간에 리뷰 요청을 주셔도 좋습니다!👏

RoadtoS7 committed:
refactor: Blackjack 게임의 프로퍼티를 private으로 변경

RoadtoS7 committed:
refactor: Blackjack.askAboutCard 메서드에서 while문의 들여쓰기 맞춤

RoadtoS7 committed:
refactor: BlackjackScore의 whos 프로퍼티 이름을 player로 수정

RoadtoS7 committed:
refactor: PlayerName 유닛테스트 케이스 이름 컨벤션을 다른 유닛테스트와 일치시킴

RoadtoS7 committed:
refactor: Player의 countScore 유닛테스트에서 XCTAssert를 다른 유닛 테스트와 일치시킴

- 수정 이유
다른 비교 테스트 구문에서는 XCTASssertEqual을 사용한 반면에 Player.countScore 테스트 코드에서만 XCTAssertTrue를 사용했었다.
비교 테스트 구문 통일 하기 위해 수정했다.

RoadtoS7 committed:
refactor: Player의 receive 유닛 테스트 이름을 구체적으로 수정함

- 수정 이유
다른 유닛 테스트 케이스에서는 메서드의 구체적인 기능 설명이 적혀있는데, Player.recive 메서드에는 적혀져 있지 않아서 수정함

RoadtoS7 committed:
fix: BlackjackScore의 whos 프로퍼티 이름을 player로 변경

RoadtoS7 committed:
refactor: Card의  CardId네이밍을 Rank로 수정

- 수정 이유
트럼프 카드에서 숫자 부분을 Rank라고 부른다. 따라서 도메인 용어에 맞춰서 Rank로 수정했다.

RoadtoS7 committed:
refactor: Card에서 사용되는 Rank, CardSymbol을 Card의 nested enum으로 수정

- 수정 이유
Card의 프로퍼티 타입으로만 사용되므로 nested type으로 수정했다.

RoadtoS7 committed:
refactor: CardSymbol 네이밍을  Suit로 수정

- 수정 이유
트럼프 카드에서 카드 문양을 영어로 Suit로 한다. 따라서 도메인 용어에 맞춰서 Suit로 수정했다

RoadtoS7 committed:
refactor: CardDeck에서 cardScoreRule 프로퍼티를 상수로 수정

- 수정 이유
초기화 이후에 변하지 않으므로 수정했다

RoadtoS7 committed:
refactor: CardPickStrategy.pick에서 뽑을 카드 집합을 Card Collection 타입으로 받도록 수정

- CardPickStrategy 입장에서는 Array, Set이든 상관없다고 판단하여 Card Collection 타입으로 수정

RoadtoS7 committed:
refactor: NonDuplicatedCardDisctributor에서 카드를 집합 자료구조로 보관하도록 수정

- 수정 이유
중복되지 않는 카드를 발행하는 CardDistributor 이므로 Set으로 카드를 보관하는 것이 더 적합하다고 판단했다

RoadtoS7 committed:
refactor: CardScoreRule의 bestScore 메서드 로직 리팩터링
dquote>
dquote> - 수정 이유
dquote> ace를 1점으로 계산한 점수와 ace를 11점으로 계산한 점수가 둘다 21점 이하인지 혹은 둘 중 하나만 21점 이하인지에 따라서 최적의 점수를 반환하도록 수정

RoadtoS7 committed:
refactor: BlackjackScoreRule의 bestScore 메서드 이름 idealScore로 수정

- 수정 이유
bestScore는 21에 가까운 점수를 반환하는 메서드인데, 가장 높은 점수를 반환하는 메서드로 인식될 수 있으므로
idealScore로 이름을 수정했다

RoadtoS7 committed:
style: BlackjackGameTest에서 들여쓰기 맞춤

RoadtoS7 committed:
style: BlackjackGameTest 의 마지막 닫는 괄호에 존재하는 공백 제거

RoadtoS7 committed:
refactor: BlackjackTest 파일 제거

- 제거 이유: 사용하지 않는다

RoadtoS7 committed:
refactor: NonDuplicateCardDistributor를 테스트할 때 테스트 대상에 카드를 한장만 두도록 수정

뽑을 카드 집합의 카드 장수가 1장일 때와 여러장일 때 테스트 케이스의 의미가 없으며
오히려 1장밖에 없을 때 카드를 뽑는 것이 더 엣지 케이에 가깝다

RoadtoS7 committed:
refactor: PlayerTest에서 countScore 유닛 테스트 이름이 어떤 테스트인지 나타나도록 수정

RoadtoS7 committed:
fix: TrumpCards가 트럼프 카드와 동일한 카드 구성을 갖고 있는지 테스트하는 로직 구현

이 유닛테스트가 미완성이었는데 완성했다

RoadtoS7 committed:
refactor: NonDuplicateCardDistributor에서 cardPickStrategy의 액세스 레벨을 private으로 수정

- cardPickStrategy 프로퍼티는 NonDuplicateCardDistributor 내부에서만 사용되므로 접근 레벨을 private으로 수정했다

zdodev commented on Blackjack/Card.swift:
```diff
@@ -0,0 +1,33 @@
+        case king
+    }
+
+    enum Suit: CaseIterable {
```
`Symbol` 도 도메인 이름으로 변경해보셨네요.👍

zdodev commented on Blackjack/CardDistributor.swift:
```diff
@@ -0,0 +1,51 @@
+        }
+        
+    }
+    private(set) var cards: Set<Card>
```
`Set` 타입으로 변경해보셨네요.👍

zdodev commented on Blackjack/CardPickStrategy.swift:
```diff
@@ -0,0 +1,20 @@
+
+
+struct RandomCardPickStrategy: CardPickStrategy {
+    func pick<Cards>(among cards: Cards) -> Card? where Cards : Collection, Cards.Element == Card {
+        return cards.randomElement()
+    }
```
깔끔하네요!👍

zdodev commented on Blackjack/CardScoreRule.swift:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+    
+    private static func idealScore(sumWhereAceOne: Int, sumWhereAceEleven: Int) -> Int {
+        let validScores = validScores(sumWhereAceOne, sumWhereAceEleven)
+        if validScores.count == 2 {
+            return validScores.max()!
+        }
+        
+        if validScores.count == 1 {
+            return validScores[0]
+        }
+        
+        return sumWhereAceOne
+    }
```
이전보다 간결해지고 파악하기 쉬워진 것 같습니다!
다만 `validScores.count == 값` 라는 표현이 아직 모호하지 않을까 싶습니다!
블랙잭 게임(도메인)을 잘 이해하고 있다면 읽기 수월할 수도 있지만
유효한 숫자가 2개이면 그 중 최대값을 반환하고, 유효한 숫자가 1개이면 그 중 첫 번째 요소를 반환한다는 로직이
블랙잭 에이스 숫자 계산법을 표현한 것과 거리가 조금 있지 않을까 싶었습니다!

다음과 같은 표현법은 어떨까요? 다음 스텝에서 고민해보면 좋을 것 같습니다!👍
```swift
return sumWhereAceEleven <= twentyOne ? sumWhereAceEleven : sumWhereAceOne
```

RoadtoS7 replied:
와 이렇게 간단하게 표현할 수 있는 로직이었군요..!! 😲
피드백 감사합니다!!

zdodev commented on BlackjackTest/TrumpCardsTest.swift:
```diff
@@ -0,0 +1,99 @@
+
+class TrumpCardsTest: XCTestCase {
+
+    func test_TrumpCards는_트럼프카드규칙대로_카드를갖는다() {
```
만약, 트럼프 카드 규칙이 바뀌어 카드가 100장이 된다면 테스트 코드도 그에 비례해서 2배씩 증가될 것 같습니다!
카드가 존재하는지 일일이 비교하는 것보다는 트럼프 카드 덱을 생성하고 카드의 총 개수가 트럼프 규칙(52장)과 동일한지 비교하는 것은 어떨까요? 이 부분은 다음 스텝에서 고민해보면 좋을 것 같습니다!👍

zdodev reviewed (approved):
안녕하세요! @RoadtoS7 

전체적으로 깔끔하고 간결하게 개선된 것 같습니다!
커밋 단위도 적절하고 커밋 내용에 수정사항 및 이유를 작성해주셔서 
의도를 파악하는데 많은 도움이 되었습니다.👍👍

질문에 대한 답변과 함께 몇 가지 코멘트를 남겨보았습니다!
다음 스텝에서 고민해보면 좋을 것 같습니다!
이번 스텝은 여기서 마무리하도록 하겠습니다. 고생하셨습니다🎉
